test/cql-pytest/test_shedding.py

```diff
@@ -31,12 +32,47 @@ def table1(cql, test_keyspace):
 # 5. Hence, a 30MiB request will be estimated to take around 60MiB RAM,
 #    which is enough to trigger shedding.
 # See also #8193.
-@pytest.mark.skip(reason="highly depends on configuration")
+#
+# Big request causes the entire connection to be closed (see #8800 for the reasons),
+# this results in NoHostAvailable on the client.
+# We check that there are no unexpected protocol_errors using Scylla Prometheus API.
+# Such errors occurred before, when before closing the connection, the remaining
+# bytes of the current request were not read to the end and were treated as
+# the beginning of the next request.
+#
+# Have no idea why, but on CI the actual memory limit for CQL requests is ~100MiB,
+# so we use ~60MiB request (which is estimated to take ~120MiB according to the logic above)
+# to guarantee shedding.
 def test_shed_too_large_request(cql, table1, scylla_only):
-    prepared = cql.prepare(f"INSERT INTO {table1} (p,t1,t2,t3,t4,t5,t6) VALUES (42,?,?,?,?,?,?)")
-    a_5mb_string = 'x'*5*1024*1024
-    with pytest.raises(InvalidRequest, match=re.compile('large', re.IGNORECASE)):
-        cql.execute(prepared, [a_5mb_string]*6)
+    def get_protocol_errors():
+        result = 0
+        for line in requests.get(f'http://{cql.cluster.contact_points[0]}:9180/metrics').text.split('\n'):
+            if not line.startswith('scylla_transport_cql_errors_total') or 'protocol_error' not in line:
+                continue
+            result += int(line.split(' ')[1])
+        return result
+
+    # protocol_errors metric is always non-zero, since the
+    # cassandra python driver use these errors to negotiate the protocol version
+    protocol_errors_before = get_protocol_errors()
+
+    # separate session is needed due to the cql_test_connection fixture,
+    # which checks that the session is not broken at the end of the test
+    with new_cql(cql) as ncql:
+        prepared = ncql.prepare(f"INSERT INTO {table1} (p,t1,t2,t3,t4,t5,t6) VALUES (42,?,?,?,?,?,?)")
+
+        # With release builds of Scylla, information that the socket is closed reaches the client driver
+        # before it has time to process the error message written by Scylla to the socket.
+        # The driver ignores unread bytes from the socket (this looks like a bug),
+        # tries to establish a connection with another node, and throws a NoHostAvailable exception if it fails.
+        # In the debug builds, the driver can have time to grab the error from the socket,
+        # and we get InvalidRequest exception.
+        with pytest.raises((NoHostAvailable, InvalidRequest),
+                           match="request size too large|Unable to complete the operation against any hosts"):
+            a_5mb_string = 'x'*10*1024*1024
+            ncql.execute(prepared, [a_5mb_string]*6)
+    protocol_errors_after = get_protocol_errors()
+    assert protocol_errors_after == protocol_errors_before
```

Comment:
The downside of tests using metrics like this is that they won't work correctly if something else is running on the same Scylla in parallel (other tests, or something else in case this is a long-living Scylla installation and not just Scylla ran for the test).
But I'll let this slide, because I think we have other tests already making the same problematic assumption :-(

Reply:
Isn't that an invitation to flakiness and "it works on my machine" reports?

Can we mark that scylladb instance to be dedicated to the test?

Reply:
@avikivity the way we normally run test/cql-pytest - through test.py or test/cql-pytest/run, always runs the tests on a dedicated Scylla instance, so it doesn't have this problem.
My comment is about the theoretical (and rarely used) option to take an already running instance of Scylla - e.g., imagine a long-running instance of Scylla on the cloud, or even another non-Scylla (competitor) CQL - and run tests against it. It can be useful to quickly diagnose if this instance is susceptible to a certain bug or not.
But as I said, this is not the way we normally run tests, and I believe we have other tests that make the assumption that tests are not run on the same Scylla instance where other stuff is happening in parallel, which is why I said I was ready to let this slide.

Reply:
ok. So now my question is why we don't run tests on shared instances, but that's not a question for this patch.

Reply:
@avikivity we can, 99% of the Alternator and cql-pytest tests are written to allow doing this - the tests use random table names, and random keys in a shared table, to allow multiple tests to use even the same table concurrently. I would like to fix the tests that don't work this way (and this test is an example of a test that would have such a problem). But I already had enough of a "battle" to convince people that it's ok that we run tests one after another on the same Scylla and even the same table created once (instead of restarting Scylla after every test as we did in dtest, boost tests, etc.) - people were afraid that it would make tests untrustable and/or undebuggible, and I didn't want to take it further to running tests in parallel on the same Scylla :-) Not to mention that Alternator tests now finish sequentially in 20 seconds, I thought it was fast enough ;-)

But as I noted, sometimes you may want to run a Scylla test in parallel to some other workload you can't control. This is the natural state of affairs on DynamoDB, for example - I can't start a DynamoDB instance just for a test.

Reply:
topology tests share instances already. these are just the same cql-pytest tests, just more powerful

Reply:
> The downside of tests using metrics like this is that they won't work correctly if something else is running on the same Scylla in parallel (other tests, or something else in case this is a long-living Scylla installation and not just Scylla ran for the test).
> But I'll let this slide, because I think we have other tests already making the same problematic assumption :-(

1. Not all tests can be run in all environments, a simple example being tests with error injection. This is just another example.
2. I would be happy to refrain from using a global Scylla state in the test, but checking the metrics was my only idea  of how to verify the behaviour we need here.